hubble-proxy/cmd/completion/completion.go

```diff
@@ -0,0 +1,106 @@
+// Copyright 2020 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package completion
+
+import (
+	"context"
+	"fmt"
+
+	cmdContext "github.com/cilium/cilium/hubble-proxy/cmd/context"
+
+	"github.com/spf13/cobra"
+)
+
+// New creates a new shell completion command.
+func New(ctx context.Context, cctx *cmdContext.Context) *cobra.Command {
+	return &cobra.Command{
+		Use:     "completion [shell]",
+		Short:   "Output shell completion code",
+		Long:    "Output shell completion code for bash and zsh",
+		Example: completionExample,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return runCompletion(ctx, cctx, cmd, args)
+		},
+		ValidArgs: []string{"bash", "zsh"},
+	}
+}
+
+func runCompletion(ctx context.Context, cctx *cmdContext.Context, cmd *cobra.Command, args []string) error {
+	if len(args) > 1 {
+		return fmt.Errorf("too many arguments; expected only the shell type: %s", args)
+	}
+
+	if len(args) == 0 || args[0] == "bash" {
```

Comment:
I noticed there's some subtle differences between the completion code here and the ones in `cmd/root.go`, is it reasonable to find a way to share this?

Reply:
Ah, yes it's because when one runs completion with a wrong argument (eg `hubble-proxy completion foo`), then the copyright header was being written anyway before the error was printed out and it felt weird. I can create a PR to fix this for cilium's cli as well.

Reply:
See #10558

Reply:
Thanks for the fix. Can we refactor / share this or is that painful for some reason (eg the way the viper context is passed around or something?)